src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -85,26 +83,64 @@ public Task RunAsync(ActionRunStage stage)
                 actionRunner.Stage = stage;
                 actionRunner.Condition = aStep.Condition;
 
-                var step = ExecutionContext.RegisterNestedStep(actionRunner, inputsData, location, Environment);
-
+                var step = ExecutionContext.RegisterNestedStep(actionRunner, inputsData, Environment);
                 InitializeScope(step);
 
-                location++;
+                // Add all steps to the Composite StepsRunner
+                // Follows similar logic to how JobRunner invokes the StepsRunner for job steps!
+                compositeSteps.Add(step);
             }
 
-            // Create a step that handles all the composite action steps' outputs
-            Pipelines.ActionStep cleanOutputsStep = new Pipelines.ActionStep();
-            cleanOutputsStep.ContextName = ExecutionContext.ContextName;
-            // Use the same reference type as our composite steps.
-            cleanOutputsStep.Reference = Action;
+            // All steps have the same initial ExecutionContext/
+            // We store one of these ExecutionContexts in this variable 
+            // so that we can process the outputs correctly.
+            var stepExecutionContext = compositeSteps?[0].ExecutionContext;
```

Comment:
have you considered using `ExecutionContext` instead of `compositeSteps?[0].ExecutionContext`?

Reply:
`ExecutionContext` is different than `compositeSteps?[0].ExecutionContext`

`ExecutionContext` contains the whole composite action's execution context. 

The `compositeSteps?[0].ExecutionContext` is the direct result of Root.CreateChild(..) in `RegisterNestedStep` and has the pointer to the correct `outputs` attribute in StepsContext that we need to use when we evaluate the outputs:
`DictionaryContextData actionOutputs = actionManifestManager.EvaluateCompositeOutputs(stepExecutionContext, Data.Outputs, evaluateContext);`

**tl;dr We can't use `ExecutionContext` since it doesn't contain the pointer to`ExpressionValues["outputs"]` while `compositeSteps?[0].ExecutionContext` has that pointer.**